### PR TITLE
docs: add python sdk docs

### PR DESCRIPTION
**Pull request by naren-determined**
## Description

Add documentation for Python SDK checkpoint export functionality.

## Checklist

- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234

### CODE REVIEW

File: harness/determined/common/experimental/checkpoint/_checkpoint.py [hunk 1]
```diff
@@ -24,6 +24,11 @@ class CheckpointState(enum.Enum):
 
 class Checkpoint(object):
     """
+    .. note::
+
+        The preferred path for checkpoint export APIs is
+        ``determined.experimental.client.get_checkpoint()``.
```
Comment (rb-determined-ai) [inline harness/determined/common/experimental/checkpoint/_checkpoint.py:30]: I think the wording of these "preferred path" docs is unclear to me.

Can we make them say something like:

```
A Checkpoint object is usually obtained from ``determined.experimental.client.get_checkpoint()``.
```

I also don't think that it needs to be a `.. note::` block.

File: docs/reference/api/client.txt [hunk 2]
```diff
@@ -0,0 +1,142 @@
+.. _client:
+
+################################
+ determined.experimental.client
+################################
+
+..
+   automodule: determined.experimental.client
```
Comment (rb-determined-ai) [inline docs/reference/api/client.txt:8]: I believe this is just a comment in ReST notation.

File: docs/reference/api/client.txt [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+   :member-order: bysource
+
+****************
+ ``Get Models``
+****************
+
+.. automodule:: determined.experimental.client
+   :noindex:
+   :members: get_models
+   :member-order: bysource
```
Comment (rb-determined-ai) [inline docs/reference/api/client.txt:79]: I believe all of these lines could just be something like:

```rest
************
 ``Client``
************

.. automodule:: determined.experimental.client
   :members:
   :member-order: bysource
```

I don't see a lot of value in exposing each function via a separate automodule section.

File: docs/reference/api/client.txt [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+
+****************
+ ``Checkpoint``
+****************
+
+.. autoclass:: determined.experimental.Checkpoint
```
Comment (rb-determined-ai) [inline docs/reference/api/client.txt:85]: It looks like if we run change each of these lines to:

```rest
.. autoclass:: determined.experimental.client.Checkpoint
```

sphinx will still find the objects, and it will display their names as `determined.experimental.client.Checkpoint`, which seems good

File: harness/determined/experimental/client.py [hunk 5]
```diff
@@ -34,27 +34,37 @@ def login(
...
        ``login(user=...)`` for any of those users without putting your password in your code.
+
+    Arguments:
+        master (string, optional): The URL of the Determined master.
+            If this argument is not specified, the environment variables
+            DET_MASTER and DET_MASTER_ADDR will be checked for the master URL in that order.
+        user (string, optional): The Determined username used for
+            authentication. (default: ``determined``)
+        password (string, optional): The password associated with the Determined
+            user.
```
Comment (rb-determined-ai) [inline harness/determined/experimental/client.py:67]: can you remove the word `Determined`?  That sounds a bit like it wants the name for the user whose username is `determined`.

File: harness/determined/experimental/client.py [hunk 6]
```diff
@@ -34,27 +34,37 @@ def login(
...
     """
```
Comment (rb-determined-ai) [inline harness/determined/experimental/client.py:68]: Additional arguments:

```
cert_path (string, optional): A path to a custom PEM-encoded certificate, against which to validate the master.  (default: ``None``)
cert_name (string, optional): The name of the master hostname to use during certificate validation.  Normally this is taken from the master URL, but there may be cases where the master is exposed on multiple networks that this value might need to be overridden.  (default: ``None``)
noverify: disable all TLS verification entirely.  (default: ``False``)
```

File: docs/reference/api/client.txt (hunk 3 above)
Comment (naren-determined) [reply docs/reference/api/client.txt:79]: If you do this, then each function won't have it's own place to click on and won't be exposed at the top-level Python API docs under `determined.experimental.client`. See the attached pic.
Screen Shot 2021-06-09 at 1 28 29 PM

Comment (rb-determined-ai) [reply docs/reference/api/client.txt:76]: That doesn't sound like a big deal to me.

File: harness/determined/common/experimental/experiment.py [hunk 7]
```diff
@@ -51,6 +51,10 @@ def __init__(self, raw: Any):
 
 class ExperimentReference:
     """
+    An ExperimentReference object is usually used in conjunction with
```
Comment (rb-determined-ai) [inline harness/determined/common/experimental/experiment.py:54]: "used in conjunction with" doesn't make much sense, I think you should say 'is usually obtained from' or 'is usually returned by' or something.

File: docs/how-to/use-trained-models.txt [hunk 8]
```diff
@@ -375,6 +375,6 @@ The checkpoint can then be loaded in Python with this code:
  Next Steps
```
Comment (rb-determined-ai) [inline docs/how-to/use-trained-models.txt:375]: Looks like this file has some old checkpoint export APIs embedded in it, can we update those?

**Review by rb-determined-ai — COMMENTED** [review]
Other files which seem to have the old checkpoint export APIs:

- examples/computer_vision/cifar10_pytorch_inference/model_def.py
- examples/gan/dcgan_tf_keras/export.py

File: docs/how-to/use-trained-models.txt [hunk 9]
```diff
@@ -55,16 +53,14 @@ order with respect to the specified metric.
 
 .. code:: python
 
-   from determined.experimental import Determined
+   from determined.experimental.client import get_experiment
```
Comment (rb-determined-ai) [inline docs/how-to/use-trained-models.txt:56]: can you make this `from determined.experimental import client`?

File: examples/gan/dcgan_tf_keras/export.py [hunk 10]
```diff
@@ -25,9 +25,7 @@ def generate_and_plot_images(generator: tf.keras.Sequential, noise_dim: int) ->
 
 
 def export_model(experiment_id: int, master_url: str) -> tf.keras.Model:
```
Comment (rb-determined-ai) [inline examples/gan/dcgan_tf_keras/export.py:27]: You will have to remove `master_url` as a parameter, and invoke `client.login()` somewhere (probably in main())

File: docs/how-to/use-trained-models.txt [hunk 11]
```diff
@@ -115,42 +115,46 @@ selecting checkpoints.
...
+loads it into memory in a Python process, as shown in the following
+snippet.
 
 .. code:: python
 
-   from determined.experimental import Determined
+   from determined.experimental import client
 
-   checkpoint = Determined().get_experiment(id).top_checkpoint()
+   client.login()
```
Comment (rb-determined-ai) [inline docs/how-to/use-trained-models.txt:156]: can we remove these `client.login()` lines?  They're often not necessary, and when they are necessary it's unlikely that  an empty `login()` call will work.

A better strategy is to point to the overall `client` docs right at the top of the document and let that be enough.

File: docs/how-to/use-trained-models.txt [hunk 12]
```diff
@@ -367,14 +375,14 @@ The checkpoint can then be loaded in Python with this code:
 
 .. code:: python
 
-   from determined.experimental import Checkpoint
+   from determined.experimental.client import Checkpoint
 
    model = Checkpoint.load_from_path("local_checkpoint")
```
Comment (rb-determined-ai) [inline docs/how-to/use-trained-models.txt:380]: should be `client.Checkpoint`

File: examples/README.md [hunk 13]
```diff
@@ -2,7 +2,7 @@
...
 | Tutorials | <p>[mnist_pytorch](tutorials/mnist_pytorch)<br>[fashion_mnist_tf_keras](tutorials/fashion_mnist_tf_keras)</p> | <p>MNIST<br>Fashion MNIST</p> | <p>PyTorch<br>TensorFlow (tf.keras)</p> |
-| Computer Vision | <p>[cifar10_pytorch](computer_vision/cifar10_pytorch)<br>[fasterrcnn_coco_pytorch](computer_vision/fasterrcnn_coco_pytorch)<br>[mmdetection_pytorch](computer_vision/mmdetection_pytorch)<br>[detr_coco_pytorch](computer_vision/detr_coco_pytorch)<br>[deformabledetr_coco_pytorch](computer_vision/deformabledetr_coco_pytorch)<br>[mnist_estimator](computer_vision/mnist_estimator)<br>[mnist_tf_layers](computer_vision/mnist_tf_layers)<br>[cifar10_tf_keras](computer_vision/cifar10_tf_keras)<br>[iris_tf_keras](computer_vision/iris_tf_keras)<br>[unets_tf_keras](computer_vision/unets_tf_keras)</p> | <p>CIFAR-10<br>Penn-Fudan Dataset<br>COCO<br>COCO<br>COCO<br>MNIST<br>MNIST<br>CIFAR-10<br>Iris Dataset<br>Oxford-IIIT Pet Dataset</p> | <p>PyTorch<br>PyTorch<br>PyTorch<br>PyTorch<br>PyTorch<br>TensorFlow (Estimator API)<br>TensorFlow (tf.layers via Estimator API)<br>TensorFlow (tf.keras)<br>TensorFlow (tf.keras)<br>TensorFlow (tf.keras)</p> |
```
Comment (rb-determined-ai) [inline examples/README.md:5]: um... what happened here?

Comment (naren-determined) [reply examples/README.md:5]: This is me adding the `cifar10_pytorch_inference` example to the README b/c I noticed it wasn't there.

File: docs/how-to/use-trained-models.txt [hunk 14]
```diff
@@ -14,15 +14,16 @@ This guide discusses:
 #. Storing additional user-defined metadata in a checkpoint.
 #. Using the Determined CLI to download checkpoints to disk.
 
+For specific on the checkpoint API, see
```
Comment (rb-determined-ai) [inline docs/how-to/use-trained-models.txt:17]: Reword?  Maybe something like:

"The Checkpoint Export API is a subset of the features found in ..."

**Review by rb-determined-ai — APPROVED** [review]
great work, thank you!
